File: app/src/main/java/com/example/volnabledemo/data/ble/AdvertisementPacketParser.kt
package com.example.volnabledemo.data.ble

import com.example.volnabledemo.domain.model.AdvertisementPacket
import com.example.volnabledemo.domain.model.VolnaContract

class AdvertisementPacketParser(
    private val qrcIdConverter: VolnaQrcIdConverter = VolnaQrcIdConverter(),
) {
    fun parse(serviceData: ByteArray): Result<AdvertisementPacket> = runCatching {
        BleLog.d("BLE_Parser", "=== Parsing serviceData ===")
        BleLog.d("BLE_Parser", "ServiceData size: ${serviceData.size}")
        BleLog.d("BLE_Parser", "ServiceData hex: ${serviceData.joinToString("") { "%02x".format(it) }}")

        require(serviceData.size >= 3) {
            "Service data too short: need at least 3 bytes, got ${serviceData.size}"
        }

        val versionAndDelta = serviceData[0].toInt() and 0xFF
        val packetVersion = (versionAndDelta shr 5) and 0b111
        BleLog.d("BLE_Parser", "Packet version: $packetVersion, expected: ${VolnaContract.supportedPacketVersion}")
        require(packetVersion == VolnaContract.supportedPacketVersion) {
            "Unsupported packet version: $packetVersion"
        }

        val deltaBits = versionAndDelta and 0b1_1111
        val rssiDelta = if ((deltaBits and 0b1_0000) != 0) deltaBits - 0b100_000 else deltaBits
        BleLog.d("BLE_Parser", "RSSI delta: $rssiDelta")

        val terminalCapabilities = serviceData[1].toInt() and 0xFF
        BleLog.d("BLE_Parser", "Terminal capabilities: 0x${terminalCapabilities.toString(16)}")
        require((terminalCapabilities and VolnaContract.requiredOnlineC2bCapabilityMask) == VolnaContract.requiredOnlineC2bCapabilityMask) {
            "Online C2B capability not supported: 0x${terminalCapabilities.toString(16)}"
        }

        val operationCounter = serviceData[2].toInt() and 0xFF
        BleLog.d("BLE_Parser", "Operation counter: $operationCounter")

        val qrcIdBytes = if (serviceData.size > 3) {
            serviceData.copyOfRange(3, serviceData.size)
        } else {
            byteArrayOf()
        }
        BleLog.d("BLE_Parser", "QR ID bytes size: ${qrcIdBytes.size}")

        val qrcId = qrcIdConverter.fromBinary(qrcIdBytes)
        BleLog.d("BLE_Parser", "QR ID: $qrcId")
        BleLog.d("BLE_Parser", "=== Parse SUCCESS ===")

        AdvertisementPacket(
            packetVersion = packetVersion,
            rssiDelta = rssiDelta,
            terminalCapabilities = terminalCapabilities,
            operationCounter = operationCounter,
            qrcId = qrcId,
        )
    }
}

File: ios/BLEApp/BLEApp/BLE/Parser/AdvertisementPacketParser.swift
import Foundation

struct AdvertisementPacket {
    let packetVersion: UInt8
    let rssiDelta: Int
    let capabilities: UInt8
    let operationCounter: UInt8
    let qrcID: String
    let qrcPayload: Data
}

enum VolnaParserError: Error, Equatable {
    case serviceDataTooShort(Int)
    case unsupportedPacketVersion(UInt8)
    case missingRequiredCapability(UInt8)
    case manufacturerDataTooShort(Int)
    case unsupportedManufacturerID(UInt16)
    case amountMustBePositive
}

struct AdvertisementPacketParser {
    private let converter = VolnaQrcIdConverter()

    func parse(_ data: Data) throws -> AdvertisementPacket {
        guard data.count >= 3 else { throw VolnaParserError.serviceDataTooShort(data.count) }

        let versionAndDelta = data[0]
        let packetVersion = (versionAndDelta >> 5) & 0b111
        guard packetVersion == BleConfig.supportedPacketVersion else {
            throw VolnaParserError.unsupportedPacketVersion(packetVersion)
        }

        let deltaBits = Int(versionAndDelta & 0b1_1111)
        let rssiDelta = (deltaBits & 0b1_0000) != 0 ? deltaBits - 0b10_0000 : deltaBits

        let capabilities = data[1]
        guard (capabilities & BleConfig.requiredCapabilityMask) == BleConfig.requiredCapabilityMask else {
            throw VolnaParserError.missingRequiredCapability(capabilities)
        }

        let operationCounter = data[2]
        let qrcPayload = data.count > 3 ? Data(data.dropFirst(3)) : Data()
        let qrcID = converter.fromBinary(qrcPayload)

        return AdvertisementPacket(
            packetVersion: packetVersion,
            rssiDelta: rssiDelta,
            capabilities: capabilities,
            operationCounter: operationCounter,
            qrcID: qrcID,
            qrcPayload: qrcPayload
        )
    }
}
